torch/csrc/distributed/c10d/NCCLUtils.hpp

```diff
@@ -179,12 +179,20 @@ std::string getNcclErrorDetailStr(
 // Write NCCL debug info to local disk or any storage users define.
 class TORCH_API DebugInfoWriter {
  public:
-  DebugInfoWriter(int rank);
   virtual ~DebugInfoWriter();
   virtual void write(const std::string& ncclTrace);
+  static DebugInfoWriter& getWriter(int rank);
+  static void registerWriter(std::unique_ptr<DebugInfoWriter> writer);
 
  protected:
+  DebugInfoWriter(std::string namePrefix, int rank) {
+    filename_ = c10::str(namePrefix, rank);
+  }
   std::string filename_;
+
+ private:
+  static std::unique_ptr<DebugInfoWriter> writer_;
+  static bool hasWriterRegistered_;
```

Comment:
Well.... I think about that. Actually we only allow register once, maybe call_once is a better choice?

Reply:
i think technically, ke is right. if 2 threads called register, both of them could race through the code and both register/set the variables.

It would be uncommon for this to happen, since most of the time there is one python process per trainer and there would be GIL protecting multiple accesses from the python side anyway.  From the c++ side, there is only rank0's watchdog that calls dump API and implicitly triggers setting the default writer.  There could still be a race between c++ setting default writer and python setting custom writer.

Reply:
On c++ side, we have a static mutex to ensure the write is serialized, so I guess that should be fine? And for python, I agree, it's less likely that more than one thread is calling it. Well, user set customized writer in python before training, so there should not have the race. (cpp registration is lazy until timeout dump)

Reply:
you could add a mutex. or you could use an atomic bool and use a CAS operation inside the register function, ensuring only the thread that changed the value of the bool is allowed to modify the writer

Reply:
OK, let me make it an atomic bool for now and we can always change it down the road.